stan/math/prim/fun/sign.hpp

```diff
@@ -2,14 +2,47 @@
 #define STAN_MATH_PRIM_FUN_SIGN_HPP
 
 #include <stan/math/prim/meta.hpp>
+#include <stan/math/prim/functor/apply_scalar_unary.hpp>
+
 namespace stan {
 namespace math {
 
 // returns 1 if NaN is passed in.
-template <typename T>
+template <typename T, require_stan_scalar_t<T>* = nullptr>
 inline int sign(const T& z) {
   return (z == 0) ? 0 : z < 0 ? -1 : 1;
 }
+
+/**
+ * Structure to wrap `sign()` so it can be vectorized.
+ */
+struct sign_fun {
+  /**
+   * Return the sign of the specified argument.
+   *
+   * @tparam T type of argument
+   * @param x argument
+   * @return sign of the argument.
+   */
+  template <typename T>
+  static inline int fun(const T& x) {
+    return sign(x);
+  }
+};
+
+/**
+ * Return the elementwise application of `sign()` to
+ * specified argument container.
+ *
+ * @tparam T type of container
+ * @param x container
+ * @return Elementwise sign of members of container.
+ */
+template <typename T, require_container_t<T>* = nullptr>
+inline auto sign(const T& x) {
+  return apply_scalar_unary<sign_fun, T>::apply(x);
+}
+
```

Comment:
I think we should go ahead and add the other generalization too: https://github.com/stan-dev/math/blob/develop/stan/math/prim/fun/acos.hpp#L58

Reply:
Eigen implements sign as `(x > 0) - (x < 0)`, which returns 0 for a NaN input. Our implementation is `(z == 0) ? 0 : z < 0 ? -1 : 1`, which returns 1 for a NaN input. We also have tests that expect result of 1 for NaN input.

Reply:
Ouch, got it.

Reply:
Wait doesn't this lead to a problem -- If this is used in a place where we have primitive Eigen types, we'll get the Eigen sign, but if we use this with vars, we'll get our sign which is different?

Reply:
Not really, unless we do `using std::sign;` or even using `Eigen::sign;`. With no using, the version from `stan::math` namespace is called.